File: totaleconomy-common/src/test/java/com/ericgrandt/totaleconomy/common/econ/CommonEconomyTest.java
package com.ericgrandt.totaleconomy.common.econ;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.ericgrandt.totaleconomy.common.data.AccountData;
import com.ericgrandt.totaleconomy.common.data.BalanceData;
import com.ericgrandt.totaleconomy.common.data.CurrencyData;
import com.ericgrandt.totaleconomy.common.domain.Account;
import com.ericgrandt.totaleconomy.common.domain.Currency;
import com.ericgrandt.totaleconomy.common.dto.CurrencyDto;
import com.ericgrandt.totaleconomy.common.logger.CommonLogger;
import java.math.BigDecimal;
import java.sql.SQLException;
import java.util.UUID;
import net.kyori.adventure.text.Component;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
public class CommonEconomyTest {
    @Mock
    private CommonLogger loggerMock;

    @Mock
    private AccountData accountDataMock;

    @Mock
    private BalanceData balanceDataMock;

    @Mock
    private CurrencyData currencyDataMock;

    @Test
    @Tag("Unit")
    public void getDefaultCurrency_WithDefaultCurrency_ShouldReturnCurrency() throws SQLException {
        // Arrange
        Currency defaultCurrency = new Currency(
            1,
            "singular",
            "plural",
            "$",
            2,
            true
        );

        when(currencyDataMock.getDefaultCurrency()).thenReturn(defaultCurrency);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        CurrencyDto actual = sut.getDefaultCurrency();
        CurrencyDto expected = new CurrencyDto(
            defaultCurrency.id(),
            defaultCurrency.nameSingular(),
            defaultCurrency.namePlural(),
            defaultCurrency.symbol(),
            defaultCurrency.numFractionDigits(),
            defaultCurrency.isDefault()
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void getDefaultCurrency_WithSqlException_ShouldReturnNull() throws SQLException {
        // Arrange
        when(currencyDataMock.getDefaultCurrency()).thenThrow(SQLException.class);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        CurrencyDto actual = sut.getDefaultCurrency();

        // Assert
        assertNull(actual);
        verify(loggerMock, times(1)).error(any(String.class), any(SQLException.class));
    }

    @Test
    @Tag("Unit")
    public void createAccount_WithSuccess_ShouldReturnTrue() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();

        when(accountDataMock.createAccount(uuid)).thenReturn(true);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        boolean actual = sut.createAccount(uuid);

        // Assert
        assertTrue(actual);
    }

    @Test
    @Tag("Unit")
    public void createAccount_WithSqlException_ShouldReturnFalse() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();

        when(accountDataMock.createAccount(uuid)).thenThrow(SQLException.class);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        boolean actual = sut.createAccount(uuid);

        // Assert
        assertFalse(actual);
        verify(loggerMock, times(1)).error(any(String.class), any(SQLException.class));
    }

    @Test
    @Tag("Unit")
    public void hasAccount_WithNoAccount_ShouldReturnFalse() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();

        when(accountDataMock.getAccount(uuid)).thenReturn(null);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        boolean actual = sut.hasAccount(uuid);

        // Assert
        assertFalse(actual);
    }

    @Test
    @Tag("Unit")
    public void hasAccount_WithSqlException_ShouldReturnFalse() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();

        when(accountDataMock.getAccount(uuid)).thenThrow(SQLException.class);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        boolean actual = sut.hasAccount(uuid);

        // Assert
        assertFalse(actual);
        verify(loggerMock, times(1)).error(any(String.class), any(SQLException.class));
    }

    @Test
    @Tag("Unit")
    public void withdraw_WithSuccess_ShouldReturnSuccessfulTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.TEN);
        when(balanceDataMock.updateBalance(uuid, currencyId, BigDecimal.ZERO)).thenReturn(1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.withdraw(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.SUCCESS,
            ""
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void withdraw_WithAmountEqualToZero_ShouldReturnFailedTransactionResult() {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.ZERO;

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.withdraw(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void withdraw_WithAmountLessThanZero_ShouldReturnFailedTransactionResult() {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.valueOf(-1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.withdraw(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void withdraw_WithInsufficientFunds_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.ONE);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.withdraw(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Insufficient funds"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void withdraw_WithSqlException_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.TEN);
        when(balanceDataMock.updateBalance(uuid, currencyId, BigDecimal.ZERO)).thenThrow(SQLException.class);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.withdraw(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "An error occurred. Please contact an administrator."
        );

        // Assert
        assertEquals(expected, actual);
        verify(loggerMock, times(1)).error(any(String.class), any(SQLException.class));
    }

    @Test
    @Tag("Unit")
    public void withdraw_WithAmountOfZeroAndAllowZeroTrue_ShouldReturnSuccessfulTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.ZERO;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.ZERO);
        when(balanceDataMock.updateBalance(uuid, currencyId, BigDecimal.ZERO)).thenReturn(1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.withdraw(uuid, currencyId, amount, true);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.SUCCESS,
            ""
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void withdraw_WithAmountLessThanZeroAndAllowZeroTrue_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.valueOf(-1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.withdraw(uuid, currencyId, amount, true);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void deposit_WithSuccess_ShouldReturnSuccessfulTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.TEN);
        when(balanceDataMock.updateBalance(uuid, currencyId, BigDecimal.valueOf(20))).thenReturn(1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.deposit(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.SUCCESS,
            ""
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void deposit_WithAmountEqualToZero_ShouldReturnFailedTransactionResult() {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.ZERO;

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.deposit(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void deposit_WithAmountLessThanZero_ShouldReturnFailedTransactionResult() {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.valueOf(-1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.deposit(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void deposit_WithSqlException_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.TEN);
        when(balanceDataMock.updateBalance(uuid, currencyId, BigDecimal.valueOf(20))).thenThrow(SQLException.class);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.deposit(uuid, currencyId, amount, false);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "An error occurred. Please contact an administrator."
        );

        // Assert
        assertEquals(expected, actual);
        verify(loggerMock, times(1)).error(any(String.class), any(SQLException.class));
    }

    @Test
    @Tag("Unit")
    public void deposit_WithAmountOfZeroAndAllowZeroTrue_ShouldReturnSuccessfulTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.ZERO;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.ZERO);
        when(balanceDataMock.updateBalance(uuid, currencyId, BigDecimal.ZERO)).thenReturn(1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.deposit(uuid, currencyId, amount, true);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.SUCCESS,
            ""
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void deposit_WithAmountLessThanZeroAndAllowZeroTrue_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.valueOf(-1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.deposit(uuid, currencyId, amount, true);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void transfer_WithSuccess_ShouldReturnTrue() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        UUID toUuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        Account toAccount = new Account(toUuid.toString(), null);

        when(accountDataMock.getAccount(toUuid)).thenReturn(toAccount);
        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.TEN);
        doNothing().when(balanceDataMock).transfer(uuid, toUuid, currencyId, amount);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.transfer(uuid, toUuid, currencyId, amount);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.SUCCESS,
            ""
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void transfer_WithAmountEqualToZero_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        UUID toUuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.ZERO;

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.transfer(uuid, toUuid, currencyId, amount);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void transfer_WithAmountLessThanZero_ShouldReturnFailedTransactionResult() {
        // Arrange
        UUID uuid = UUID.randomUUID();
        UUID toUuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.valueOf(-1);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.transfer(uuid, toUuid, currencyId, amount);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Invalid amount"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void transfer_WithNoAccountForToUuid_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        UUID toUuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        when(accountDataMock.getAccount(toUuid)).thenReturn(null);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.transfer(uuid, toUuid, currencyId, amount);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "User not found"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void transfer_WithInsufficientFunds_ShouldReturnTrue() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        UUID toUuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        Account toAccount = new Account(toUuid.toString(), null);

        when(accountDataMock.getAccount(toUuid)).thenReturn(toAccount);
        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.ONE);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.transfer(uuid, toUuid, currencyId, amount);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "Insufficient funds"
        );

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void transfer_WithSqlException_ShouldReturnFailedTransactionResult() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        UUID toUuid = UUID.randomUUID();
        int currencyId = 1;
        BigDecimal amount = BigDecimal.TEN;

        Account toAccount = new Account(toUuid.toString(), null);

        when(accountDataMock.getAccount(toUuid)).thenReturn(toAccount);
        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.TEN);
        doThrow(SQLException.class).when(balanceDataMock).transfer(uuid, toUuid, currencyId, amount);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        TransactionResult actual = sut.transfer(uuid, toUuid, currencyId, amount);
        TransactionResult expected = new TransactionResult(
            TransactionResult.ResultType.FAILURE,
            "An error occurred. Please contact an administrator."
        );

        // Assert
        assertEquals(expected, actual);
        verify(loggerMock, times(1)).error(any(String.class), any(SQLException.class));
    }

    @Test
    @Tag("Unit")
    public void getBalance_WithBalance_ShouldReturnBalance() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenReturn(BigDecimal.TEN);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        BigDecimal actual = sut.getBalance(uuid, currencyId);
        BigDecimal expected = BigDecimal.TEN;

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void getBalance_WithSqlException_ShouldReturnNull() throws SQLException {
        // Arrange
        UUID uuid = UUID.randomUUID();
        int currencyId = 1;

        when(balanceDataMock.getBalance(uuid, currencyId)).thenThrow(SQLException.class);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        BigDecimal actual = sut.getBalance(uuid, currencyId);
        BigDecimal expected = BigDecimal.ZERO;

        // Assert
        assertEquals(expected, actual);
        verify(loggerMock, times(1)).error(any(String.class), any(SQLException.class));
    }

    @Test
    @Tag("Unit")
    public void format_WithFractionalDigits_ShouldReturnFormattedAmountWithCorrectDecimalPlaces() {
        // Arrange
        CurrencyDto currencyDto = new CurrencyDto(
            1,
            "Dollar",
            "Dollars",
            "$",
            2,
            true
        );
        BigDecimal amount = BigDecimal.valueOf(10.1234);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        Component actual = sut.format(currencyDto, amount);
        Component expected = Component.text("$10.12");

        // Assert
        assertEquals(expected, actual);
    }

    @Test
    @Tag("Unit")
    public void format_WithNoFractionalDigits_ShouldReturnFormattedAmountWithNoDecimalPlaces() {
        // Arrange
        CurrencyDto currencyDto = new CurrencyDto(
            1,
            "Dollar",
            "Dollars",
            "$",
            0,
            true
        );
        BigDecimal amount = BigDecimal.valueOf(10.1234);

        CommonEconomy sut = new CommonEconomy(loggerMock, accountDataMock, balanceDataMock, currencyDataMock);

        // Act
        Component actual = sut.format(currencyDto, amount);
        Component expected = Component.text("$10");

        // Assert
        assertEquals(expected, actual);
    }
}

File: totaleconomy-common/src/test/kotlin/com/ericgrandt/totaleconomy/econ/CommonEconomyTest.kt
package com.ericgrandt.totaleconomy.econ

import com.ericgrandt.totaleconomy.data.AccountData
import com.ericgrandt.totaleconomy.data.BalanceData
import com.ericgrandt.totaleconomy.data.entity.Account
import com.ericgrandt.totaleconomy.data.entity.Balance
import com.ericgrandt.totaleconomy.model.DatabaseError
import com.ericgrandt.totaleconomy.model.DatabaseErrorN
import com.ericgrandt.totaleconomy.model.DepositIntoBalance
import com.ericgrandt.totaleconomy.model.SetBalance
import com.ericgrandt.totaleconomy.model.WithdrawFromBalance
import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Ok
import com.ericgrandt.totaleconomy.result.Err as ErrOld
import com.ericgrandt.totaleconomy.result.Ok as OkOld
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.mockk
import net.kyori.adventure.text.Component
import java.sql.SQLException
import java.util.UUID
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

class CommonEconomyTest {
    @MockK
    lateinit var accountDataMock: AccountData

    @MockK
    lateinit var balanceDataMock: BalanceData

    @BeforeTest
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    fun currencyNamePlural_ShouldReturnCorrectString() {
        // Arrange
        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.currencyNamePlural()
        val expected = "Diamonds"

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun currencyNameSingular_ShouldReturnCorrectString() {
        // Arrange
        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.currencyNameSingular()
        val expected = "Diamond"

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun createAccount_WithSuccessResultFromAccountData_ShouldReturnTrue() {
        // Arrange
        every { accountDataMock.createAccount(any()) } returns Ok(1)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.createAccount(UUID.randomUUID())
        val expected = Ok(1)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun createAccount_WithErrorFromAccountData_ShouldReturnDatabaseError() {
        // Arrange
        every { accountDataMock.createAccount(any()) } returns Err(SQLException())

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.createAccount(UUID.randomUUID())
        val expected = Err(DatabaseError)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun hasAccount_WithAnAccount_ShouldReturnTrue() {
        // Arrange
        val accountMock = mockk<Account>()
        every { accountDataMock.getAccount(any()) } returns Ok(accountMock)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.hasAccount(UUID.randomUUID())
        val expected = Ok(true)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun hasAccount_WithNoAccount_ShouldReturnFalse() {
        // Arrange
        every { accountDataMock.getAccount(any()) } returns Ok(null)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.hasAccount(UUID.randomUUID())
        val expected = Ok(false)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun hasAccount_WithErrorGettingAccount_ShouldReturnDatabaseError() {
        // Arrange
        every { accountDataMock.getAccount(any()) } returns Err(SQLException())

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.hasAccount(UUID.randomUUID())
        val expected = Err(DatabaseError)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun getBalance_WithBalance_ShouldReturnBalance() {
        // Arrange
        val balanceMock = mockk<Balance>()
        every { balanceMock.balance } returns 1.23
        every { balanceDataMock.getBalance(any()) } returns OkOld(balanceMock)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.getBalance(UUID.randomUUID())
        val expected = OkOld(1.23)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun getBalance_WithNullBalance_ShouldReturnZeroBalance() {
        // Arrange
        every { balanceDataMock.getBalance(any()) } returns OkOld(null)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.getBalance(UUID.randomUUID())
        val expected = OkOld(0.00)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun getBalance_WithErrorGettingBalance_ShouldReturnDatabaseError() {
        // Arrange
        every { balanceDataMock.getBalance(any()) } returns ErrOld(SQLException())

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.getBalance(UUID.randomUUID())
        val expected = ErrOld(DatabaseErrorN)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun setBalance_WithSuccess_ShouldReturnNumberOfUpdatedRows() {
        // Arrange
        every { balanceDataMock.setBalance(any()) } returns OkOld(1)

        val input = SetBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.setBalance(input)
        val expected = OkOld(1)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun setBalance_WithErrorGettingBalance_ShouldReturnDatabaseError() {
        // Arrange
        every { balanceDataMock.setBalance(any()) } returns ErrOld(SQLException())

        val input = SetBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.setBalance(input)
        val expected = ErrOld(DatabaseErrorN)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun withdrawFromBalance_WithSuccess_ShouldReturnUpdatedBalance() {
        // Arrange
        val mockBalance = Balance(UUID.randomUUID(), UUID.randomUUID(), 1.0);

        every { balanceDataMock.withdrawFromBalance(any()) } returns OkOld(1)
        every { balanceDataMock.getBalance(any()) } returns OkOld(mockBalance)

        val input = WithdrawFromBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.withdrawFromBalance(input)
        val expected = OkOld(mockBalance)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun withdrawFromBalance_WithErrorWithdrawing_ShouldReturnDatabaseError() {
        // Arrange
        val mockBalance = Balance(UUID.randomUUID(), UUID.randomUUID(), 1.0);

        every { balanceDataMock.withdrawFromBalance(any()) } returns ErrOld(SQLException())
        every { balanceDataMock.getBalance(any()) } returns OkOld(mockBalance)

        val input = WithdrawFromBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.withdrawFromBalance(input)
        val expected = ErrOld(DatabaseErrorN)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun withdrawFromBalance_WithErrorGettingBalanceAfterWithdraw_ShouldReturnDatabaseError() {
        // Arrange
        every { balanceDataMock.withdrawFromBalance(any()) } returns OkOld(1)
        every { balanceDataMock.getBalance(any()) } returns ErrOld(SQLException())

        val input = WithdrawFromBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.withdrawFromBalance(input)
        val expected = ErrOld(DatabaseErrorN)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun depositIntoBalance_WithSuccess_ShouldReturnUpdatedBalance() {
        // Arrange
        val mockBalance = Balance(UUID.randomUUID(), UUID.randomUUID(), 1.0);

        every { balanceDataMock.depositIntoBalance(any()) } returns OkOld(1)
        every { balanceDataMock.getBalance(any()) } returns OkOld(mockBalance)

        val input = DepositIntoBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.depositIntoBalance(input)
        val expected = OkOld(mockBalance)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun depositIntoBalance_WithErrorDepositing_ShouldReturnDatabaseError() {
        // Arrange
        val mockBalance = Balance(UUID.randomUUID(), UUID.randomUUID(), 1.0);

        every { balanceDataMock.depositIntoBalance(any()) } returns ErrOld(SQLException())
        every { balanceDataMock.getBalance(any()) } returns OkOld(mockBalance)

        val input = DepositIntoBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.depositIntoBalance(input)
        val expected = ErrOld(DatabaseErrorN)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun depositIntoBalance_WithErrorGettingBalanceAfterDeposit_ShouldReturnDatabaseError() {
        // Arrange
        every { balanceDataMock.depositIntoBalance(any()) } returns OkOld(1)
        every { balanceDataMock.getBalance(any()) } returns ErrOld(SQLException())

        val input = DepositIntoBalance(UUID.randomUUID(), 10.0)

        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.depositIntoBalance(input)
        val expected = ErrOld(DatabaseErrorN)

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun format_WithAmountEqualToOne_ShouldReturnFormattedAmountWithSingularCurrencyName() {
        // Arrange
        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.format(1.0)
        val expected = Component.text("1.00 Diamond")

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun format_WithAmountGreaterThanOne_ShouldReturnFormattedAmountWithPluralCurrencyName() {
        // Arrange
        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.format(10.52)
        val expected = Component.text("10.52 Diamonds")

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun format_WithMoreThanTwoFractionalDigitsAndThirdNumberBeingUnderFive_ShouldRoundDown() {
        // Arrange
        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.format(10.521651234)
        val expected = Component.text("10.52 Diamonds")

        // Assert
        assertEquals(expected, actual)
    }

    @Test
    fun format_WithMoreThanTwoFractionalDigitsAndThirdNumberBeingFive_ShouldRoundUp() {
        // Arrange
        val sut = CommonEconomy(accountDataMock, balanceDataMock)

        // Act
        val actual = sut.format(10.5252)
        val expected = Component.text("10.53 Diamonds")

        // Assert
        assertEquals(expected, actual)
    }
}